x/mongo/driver/topology/connection.go

```diff
@@ -484,6 +436,12 @@ func (c *connection) read(ctx context.Context) (bytesRead []byte, errMsg string,
 }
 
 func (c *connection) close() error {
+	// Stop any blocking operations occurring in connect(), but await closing the
+	// connections directly before closing the connection context. This ensures
+	// that closing a connection will manifest as an io.EOF error, avoiding
+	// non-deterministic connection closure errors.
+	defer c.closeConnectContext()
```

Comment:
@matthewdale We must defer closing the connect context to avoid non-deterministic closure errors. For example running `TestErrors/errors_are_wrapped/network_error_during_auth`, not defering can result in 1 of 2 errors when closing a connection:

- io.EOF
- connection() error occurred during connection handshake: auth error: sasl conversation error: unable to authenticate using mechanism "SCRAM-SHA-256":  connection(localhost:27018[-1761]) socket was unexpectedly closed: context canceled: connection(localhost:27018[-1761]) socket was unexpectedly closed: context canceled